File: backend/internal/domain/todo.go
package domain

import (
	"context"
	"errors"
	"time"
)

var ErrNotFound = errors.New("not found")

const (
	TodoStatusTodo       = "todo"
	TodoStatusInProgress = "in_progress"
	TodoStatusDone       = "done"
	TodoStatusCancelled  = "cancelled"
)

var ValidStatuses = map[string]bool{
	TodoStatusTodo:       true,
	TodoStatusInProgress: true,
	TodoStatusDone:       true,
	TodoStatusCancelled:  true,
}

const (
	TodoKindNote  = "note"
	TodoKindIssue = "issue"
)

var ValidKinds = map[string]bool{
	TodoKindNote:  true,
	TodoKindIssue: true,
}

const (
	IssueTypeFeature     = "feature"
	IssueTypeBug         = "bug"
	IssueTypeImprovement = "improvement"
)

var ValidIssueTypes = map[string]bool{
	IssueTypeFeature:     true,
	IssueTypeBug:         true,
	IssueTypeImprovement: true,
}

// Todo represents a single to-do item.
type Todo struct {
	ID        int64
	Title     string
	Body      string
	Status    string
	Priority  int
	DueDate   *string
	Kind      string
	IssueType *string
	CreatedAt time.Time
	UpdatedAt time.Time
}

type HealthStatus struct {
	Status string `json:"status"`
	DBOk   bool   `json:"db_ok"`
}

// TodoStore defines the persistence contract for Todo entities.
// All concrete implementations (e.g. SQLiteStore) must satisfy this interface.
type TodoStore interface {
	Create(ctx context.Context, todo *Todo) error
	GetAll(ctx context.Context) ([]*Todo, error)
	GetByID(ctx context.Context, id int64) (*Todo, error)
	Update(ctx context.Context, todo *Todo) error
	Delete(ctx context.Context, id int64) error
}

File: frontend/src/types/todo.ts
export type TodoStatus = 'todo' | 'in_progress' | 'done' | 'cancelled';
export type Priority = 0 | 1 | 2 | 3;
export type TodoKind = 'note' | 'issue';
export type IssueType = 'feature' | 'bug' | 'improvement';

export interface Todo {
  id: number;
  title: string;
  body: string;
  status: TodoStatus;
  priority: Priority;
  due_date: string | null;
  kind: TodoKind;
  issue_type: IssueType | null;
  project_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface CreateTodoPayload {
  title: string;
  body?: string;
  priority?: Priority;
  due_date?: string | null;
  kind?: TodoKind;
  issue_type?: IssueType | null;
  project_id?: number | null;
}

export interface UpdateTodoPayload {
  title?: string;
  body?: string;
  status?: TodoStatus;
  priority?: Priority;
  due_date?: string | null;
  kind?: TodoKind;
  issue_type?: IssueType | null;
  project_id?: number | null;
}

export class ApiError extends Error {
  constructor(
    public status: number,
    message: string,
  ) {
    super(message);
    this.name = 'ApiError';
  }
}
